File: algo/73. Set Matrix Zeroes.py
# O(m+n) space, Time complexity O(mn)
# O(mn) space if keeping track the index of all zero values

class Solution:
    def setZeroes(self, matrix: 'List[List[int]]') -> 'None':
        """
        Do not return anything, modify matrix in-place instead.
        """
        numRows = len(matrix)
        numColumns = len(matrix[0])
        hasZeroRows = [0]*numRows
        hasZeroColumns = [0]*numColumns
        for row_index, row in enumerate(matrix):
            if 0 in row:
                hasZeroRows[row_index] = 1
        for column_index, column in enumerate(zip(*matrix)):
            if 0 in column:
                hasZeroColumns[column_index] = 1
                
        for row_index, hasZero in enumerate(hasZeroRows):
            if hasZero:
                matrix[row_index] = [0]*numColumns
        for column_index, hasZero in enumerate(hasZeroColumns):
            if hasZero:
                for i in range(numRows):
                    matrix[i][column_index] = 0
                

File: round2/73. Set Matrix Zeroes.cs
public class Solution {
    public void SetZeroes(int[][] matrix) {
        int m = matrix.Length, n = matrix[0].Length;
        bool rowFlag = false, colFlag = false;
        for (int i=0; i<m; i++) {
            if (matrix[i][0] == 0) {
                colFlag = true;
                break;
            }
        }
        for (int j=0; j<n; j++) {
            if (matrix[0][j] == 0) {
                rowFlag = true;
                break;
            }
        }
        for (int i=1; i<m; i++) {
            for (int j=1; j<n; j++) {
                if (matrix[i][j] == 0) {
                    matrix[i][0] = 0;
                    matrix[0][j] = 0;
                }
            }
        }
        for (int i=1; i<m; i++) {
            for (int j=1; j<n; j++) {
                if (matrix[i][0] == 0 || matrix[0][j] == 0) {
                    matrix[i][j] = 0;
                }
            }
        }
        for (int i=0; i<m; i++) {
            if (colFlag) matrix[i][0] = 0;
        }
        for (int j=0; j<n; j++) {
            if (rowFlag) matrix[0][j] = 0;
        }
    }
}